Create PERL_UNIQUE_NAME and use it to avoid name collisions

khwilliamson opened the pull request:
`PERL_UNIQUE_NAME` is a factory creating a preprocessor symbol that is based on an input skeleton name, and the line number it is called from.  The skeleton name would likely be something like `sv1_` or `PL_sv1`.  These should be chosen to not have conflicts with the XS code.  The advantage of this approach is that you won't be using the same variable name in a bunch of macros and they collide.  

* This set of changes does not require a perldelta entry.

jkeenan commented:
> Both of these make me wonder if they wouldn't be better off as an inline function.

@khwilliamson could you respond to the above?  Thanks.

khwilliamson committed:
Add macro PERL_UNIQUE_NAME

This macro acts like a factory to produce a name from its argument that
also includes the line number.

khwilliamson committed:
Change names in OpTYPE_set() to avoid possible collisions

This uses the PERL_UNIQUE_NAME() macro created in the previous commit to
get unique names for variables in macro expansions.

Suppose you have

    #define foo STMT_START { ... } STMT_END

block, and have a variable 'bar' declared in it.  By first saying

    #define bar PERL_UNIQUE_NAME(bar)

'bar' will be replaced in the expansion of 'foo' by a unique value.

It is your responsibility to not have two 'bar' symbols in the same
program.  But if you forget, you will be warned by the compiler that
'bar' is redefined, so you can fix it
.

tonycoz commented:
I noticed this when looking for the OpTYPE_set() source (looking at @scottchiefbaker's `uinteger`) - it has the same problem as was mentioned elsewhere, `o1_` and `type1_` are replaced globally, if someone is using a trailing underscore for [structure members](https://stackoverflow.com/questions/3650623/trailing-underscores-for-member-variables-in-c), this is going to replace them with line specific nonsense.

You could handle this by using two macros, something like (untested):
```
#define OpTYPE_set_(o,type, o1, type1)                      \
    STMT_START {                                \
        OP *o1 = (OP *)o;                      \
        OPCODE type1 = type;                   \
        o1_->op_type = type1;                  \
        o1_->op_ppaddr = PL_ppaddr[type1];     \
    } STMT_END

#define OpTYPE_set(o, type) OpTYPE_set_(o, type, PERL_UNIQUE_NAME(o_temp), PERL_UNIQUE_NAME(type_temp))
```

Doing this limits the visibility of the replacement to the macro body.

Or we can use an inline function.
